src/Covenant/Type.hs

```diff
@@ -588,3 +656,136 @@ returnHelper xs = case NonEmpty.uncons xs of
+          Unifiable index1 -> pure . Map.singleton index1 $ actual
+          Rigid level1 index1 -> expectRigid level1 index1
+          Wildcard scopeId1 index1 -> expectWildcard scopeId1 index1
+        ThunkT t1 -> expectThunk t1
+        BuiltinFlat t1 -> expectFlatBuiltin t1
+        BuiltinNested t1 -> case t1 of
+          ListT _ t1' -> expectListOf t1'
+          PairT t11 t12 -> expectPairOf t11 t12
+    )
+    (promoteUnificationError expected actual)
+  where
+    unificationError :: forall (a :: Type). Either TypeAppError a
+    unificationError = Left . DoesNotUnify expected $ actual
+    noSubUnify :: forall (k :: Type) (a :: Type). Either TypeAppError (Map k a)
+    noSubUnify = pure Map.empty
+    expectRigid ::
+      Int -> Index "tyvar" -> Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+    -- Rigids behave identically to concrete types: they can unify with
+    -- themselves, or any other abstraction, but nothing else. No substitutional
+    -- rewrites are needed.
+    expectRigid level1 index1 = case actual of
+      Abstraction (Rigid level2 index2) ->
+        if level1 == level2 && index1 == index2
+          then noSubUnify
+          else unificationError
+      Abstraction _ -> noSubUnify
+      _ -> unificationError
+    expectWildcard ::
+      Word64 -> Index "tyvar" -> Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+    -- Wildcards can unify with unifiables, as well as themselves, but nothing
+    -- else. No substitutional rewrites are needed.
+    expectWildcard scopeId1 index1 = case actual of
+      Abstraction (Unifiable _) -> noSubUnify
+      Abstraction (Wildcard scopeId2 index2) ->
+        if scopeId1 /= scopeId2 || index1 == index2
+          then noSubUnify
+          else unificationError
+      _ -> unificationError
+    expectThunk :: CompT Renamed -> Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+    -- Thunks unify unconditionally with wildcards or unifiables. They unify
+    -- conditionally with other thunks, provided that we can unify each argument
+    -- with its counterpart in the same position, as well as their result types,
+    -- without conflicts.
+    expectThunk (CompT _ t1) = case actual of
+      Abstraction (Rigid _ _) -> unificationError
+      Abstraction _ -> noSubUnify
+      ThunkT (CompT _ t2) -> do
+        unless (comparing NonEmpty.length t1 t2 == EQ) unificationError
+        catchError
+          (foldM (\acc (l, r) -> unify l r >>= reconcile acc) Map.empty . NonEmpty.zip t1 $ t2)
+          (promoteUnificationError expected actual)
+      _ -> unificationError
+    expectFlatBuiltin :: BuiltinFlatT -> Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+    -- 'Flat' builtins are always concrete. They can unify with themselves,
+    -- unifiables or wildcards, but nothing else. No substitutional rewrites are
+    -- needed.
+    expectFlatBuiltin t1 = case actual of
+      Abstraction (Rigid _ _) -> unificationError
```

Comment:
Why rigid can't unify with a concrete type when Rigid is a concrete type?

Reply:
Because rigids are a type we don't know. Essentially, someone in an enclosing scope chooses what the rigid is, so we can't be sure if it's a concrete type at all, or even if it is, what concrete type it is.